### PR TITLE
Detect undefined values

### PR DESCRIPTION
**Pull request by cjllanwarne**
Lets the intelliJ plugin detect and error-highlight values which haven't been declared.

eg:
screen shot 2017-09-22 at 2 07 54 pm

### CODE REVIEW

**Review by danbills — APPROVED** [review]
winstanley will miss you when you're on vaca.

(We should probably update docs to point people to Intellij rather than Sublime too.)

**Comment by katevoss** [comment]
@danbills I think this is the WDL page you're looking for: https://software.broadinstitute.org/wdl/toolkit

**Comment by cjllanwarne** [comment]
@katevoss is this something we can actually update?

I know @vdauwera originally preferred sublime since it was lighter-weight, but maybe having error detection and auto-completion in the editor could swing her round?

**Comment by vdauwera** [comment]
Maybe we just present people with both options, pros and cons, and let them pick what suits their lifestyle. Like the sleep number thing but for coding.

File: src/winstanley/structure/WdlImplicits.scala [hunk 1]
```diff
@@ -34,6 +35,42 @@ object WdlDocumentUtils {
...
+
+    def getIdentifierNode: Option[ASTNode] = psiElement.getNode.getChildren(null).collectFirst {
+      case id if id.getElementType == WdlTypes.IDENTIFIER => id
+    }
+  }
+
+  implicit final class EnhancedWdlDeclaration(val wdlDeclaration: WdlDeclaration) extends AnyVal {
+    def declaredValueName: String = wdlDeclaration.getIdentifierNode match {
+      case Some(d) => d.getText
+      case None => "whaaa"
```
Comment (mcovarr) [inline src/winstanley/structure/WdlImplicits.scala]: If this wasn't left in by mistake it at least needs a comment...

File: src/winstanley/structure/WdlImplicits.scala [hunk 2]
```diff
@@ -34,6 +35,42 @@ object WdlDocumentUtils {
...
+    }
+  }
+
+  implicit final class EnhancedWdlValue(val wdlValue: WdlValue) extends AnyVal {
+    def asIdentifierNode: Option[ASTNode] = {
+      if (wdlValue.getChildren.isEmpty) { wdlValue.getIdentifierNode } else { None }
```
Comment (mcovarr) [inline src/winstanley/structure/WdlImplicits.scala]: ToL the curlies seem superfluous

File: src/winstanley/structure/WdlImplicits.scala [hunk 3]
```diff
@@ -34,6 +35,42 @@ object WdlDocumentUtils {
         rbrace <- Option(psiElement.getNode.findChildByType(rtype))
       } yield new TextRange(lbrace.getTextRange.getStartOffset + 1, rbrace.getTextRange.getEndOffset - 1)
     }
+
+    def findContainingScatter: Option[WdlScatterBlock] = {
+      Option(psiElement.getParent) flatMap {
+        case s: WdlScatterBlock => Some(s)
```
Comment (mcovarr) [inline src/winstanley/structure/WdlImplicits.scala]: `Option`

**Review by mcovarr — APPROVED** [review]

File: src/winstanley/structure/WdlImplicits.scala (hunk 3 above)
Comment (cjllanwarne) [reply src/winstanley/structure/WdlImplicits.scala]: `SomeSafe`?

**Pull request merged**